### ISSUE TITLE
[lld] After c30e6447c02, duplicate symbol errors with dtrace-processed object files

### ISSUE THREAD

**Issue by DimitryAndric** [llvm/llvm-project#59419]
As reported in https://bugs.freebsd.org/265516, https://bugs.freebsd.org/260947 but also visible in https://pkg-status.freebsd.org/gohan05/data/maini386PR265425-default/2022-12-06_12h24m32s/logs/errors/ghc-9.2.5.log, with lld 15 several FreeBSD ports using DTrace error out with "duplicate symbol" errors, specifically for DTrace-related symbols.

Bisection shows that this started appearing after c30e6447c0225f675773d07f2b6d4e3c2b962155 ("[ELF] Move section assignment from initializeSymbols to postParse"), possibly due to the "Postpone reporting duplicate symbol errors so that the messages have the section information" item described in the commit message.

Reproduction tarball with perl 5.36 objects: [lld-dtrace-dupe.tar.gz](https://github.com/llvm/llvm-project/files/10195424/lld-dtrace-dupe.tar.gz)

@emaste reported in https://bugs.freebsd.org/258872#c4 that this may have something to do with DTrace making local symbols hidden. Still, it is strange that a seemingly internal-only change in lld causes errors that weren't there before.

**Comment by MaskRay** [comment]
The DTrace use is invalid. Multiple relocatable files define `st_value=0` absolute symbols of the same name.

```
% readelf -Ws wrkdirs/share/dim/ports/lang/perl5.36/work/perl-5.36.0/mpdtrace/perlmini.o | grep __dtraceenabled_perl___op__entry
   263: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  ABS __dtraceenabled_perl___op__entry
```

In the rigid interpretation of http://www.sco.com/developers/gabi/latest/ch4.symtab.html , `STB_GLOBAL` absolute symbols
should cause duplicate definition errors as non-absolute symbols. In reality, for some reason GNU ld has code to allow
`STB_GLOBAL` absolute symbols of the same `st_value`. I am on the fence of this behavior: I don't like it deviates from
the standard but feel like this extension does make sense to some extent. lld copied this hack.
c30e6447c0225f675773d07f2b6d4e3c2b962155 copied the hack but dropped the support for `st_value=0` to efficiently detect
bitcode files (which don't allow the hack). I think c30e6447c0225f675773d07f2b6d4e3c2b962155 made a pretty reasonable
compatibility trade-off.

I think DTrace should switch to `STB_WEAK SHN_ABS` if it wants to keep using `st_value=0`.

**Comment by DimitryAndric** [comment]
Ping @markjdb, since you're the DTrace person, any remarks? I don't know enough about this area...

**Comment by markjdb** [comment]
I tried changing dtrace to modify the probe symbols to be `STB_WEAK` symbols. It seems to fix perl 5.36 and doesn't break the USDT regression tests, and I can't see why it would cause any problems. I'll have to do some more testing though.

The original dtrace code set the section index to SHN_SUNW_IGNORE, which presumably is specific to Sun's linker. I changed it to set SHN_ABS since I discovered that the aforementioned GNU ld hack works.

So I suspect that this issue can be closed, i.e., we'll fix the problem in dtrace.

> @emaste reported in https://bugs.freebsd.org/258872#c4 that this may have something to do with DTrace making local symbols hidden.

If I understand the problem correctly, this is unrelated.
